### PR TITLE
CollapseMultipleEdges

### PR DESCRIPTION
**Pull request by marinaanagno**

### CODE REVIEW

**Comment by samanthaharper** [comment]
An example of the difference in time between the two functions. 

`gap> D := RandomMultiDigraph(10000);
<immutable multidigraph with 10000 vertices, 35859826 edges>
gap> DIGRAPHS_CollapseMultipleEdges(D);; time;
24409`

`gap> DIGRAPHS_CollapseMultiColouredEdges(D, fail);; time;
105996`

**Review by le27 — APPROVED** [review]
Everything seems good to me :)

File: gap/isomorph.gi [hunk 1]
```diff
@@ -651,64 +655,89 @@ end);
...
-    if Length(adjv) > 0 then
-      indices := [idx .. idx + Length(adjv) - 1];
-      colsv := ShallowCopy(edge_colours[v]);
-      p := Sortex(adjv);
-      colsv := Permuted(colsv, p);
-      indices := Permuted(indices, p);
+    nbs := OutNeighbours(D)[v];
+    if not IsEmpty(nbs) then
+      adjv := ShallowCopy(nbs);
+      if Length(adjv) > 0 then
```
[the remark this answers is not in the document]
Comment (marinaanagno) [reply gap/isomorph.gi]: Thanks @le27 just deleted the unnecessary if statement :)

**Comment by james-d-mitchell** [comment]
Resolves Issue #306

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
A couple of minor things to fix, then you're done!

**Comment by james-d-mitchell** [comment]
Looks like the CI is failing because something in the C code for clique finding is broken. I.e. this PR is not responsible for the failing tests.

File: gap/isomorph.gi [hunk 2]
```diff
@@ -651,29 +655,26 @@ end);
 InstallGlobalFunction(DIGRAPHS_CollapseMultiColouredEdges,
 function(D, edge_colours)
   local n, mults, out, new_cols, new_cols_set, idx, adjv, indices, colsv,
-  p, run, cur, range, cols, C, v, i;
+  p, run, cur, range, cols, C, v, i, nbs;
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:658]:
```suggestion
  p, run, cur, range, cols, v, i, nbs;
```

File: gap/isomorph.gi [hunk 3]
```diff
@@ -651,64 +655,87 @@ end);
...
+            AddSet(new_cols_set, cols);
+            Add(new_cols[v], cols);
+            run := 1;
+          fi;
+          cur := cur + 1;
+        od;
+      idx := idx + Length(nbs);
+    fi;
+  od;
+    for cols in new_cols do
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply gap/isomorph.gi:698]:
```suggestion
  for cols in new_cols do
```

File: gap/isomorph.gi [hunk 4]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
-      AddSet(new_cols_set, cols);
-      Add(new_cols[v], cols);
+      idx := idx + Length(nbs);
     fi;
-    idx := idx + Length(adjv);
   od;
   for cols in new_cols do
-    for i in [1 .. Length(cols)] do
-      cols[i] := Position(new_cols_set, cols[i]);
+      for i in [1 .. Length(cols)] do
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:699]:
```suggestion
    for i in [1 .. Length(cols)] do
```

File: gap/isomorph.gi [hunk 5]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
+        cols[i] := Position(new_cols_set, cols[i]);
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:700]:
```suggestion
      cols[i] := Position(new_cols_set, cols[i]);
```

File: gap/isomorph.gi [hunk 6]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
+      od;
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:701]:
```suggestion
    od;
```

File: gap/isomorph.gi [hunk 7]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
     od;
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:702]:
```suggestion
  od;
```

File: gap/isomorph.gi [hunk 8]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
+    return [Digraph(out), new_cols, mults];
+  end);
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:704]:
```suggestion
end);
```

File: gap/isomorph.gi [hunk 9]
```diff
@@ -691,24 +692,48 @@ function(D, edge_colours)
...
-    idx := idx + Length(adjv);
   od;
   for cols in new_cols do
-    for i in [1 .. Length(cols)] do
-      cols[i] := Position(new_cols_set, cols[i]);
+      for i in [1 .. Length(cols)] do
+        cols[i] := Position(new_cols_set, cols[i]);
+      od;
     od;
+    return [Digraph(out), new_cols, mults];
```
Comment (james-d-mitchell) [inline gap/isomorph.gi:703]:
```suggestion
  return [Digraph(out), new_cols, mults];
```

**Review by james-d-mitchell — DISMISSED** [review]
There were a couple of minor things to fix, over indented lines etc, I just committed fixes for those, and otherwise I am happy with this.

**Comment by marinaanagno** [comment]
Thanks for reviewing this @james-d-mitchell !

**Comment by wilfwilson** [comment]
Does this have some usefulness/relation to `DigraphRemoveAllMultipleEdges`? Should `DigraphRemoveAllMultipleEdges` perhaps use this new function as its backend? Just an idea, haven't thought about it beyond writing this sentence.

**Review by james-d-mitchell — APPROVED** [review]
Looks good to me!

**Comment by james-d-mitchell** [comment]
> Does this have some usefulness/relation to `DigraphRemoveAllMultipleEdges`? Should `DigraphRemoveAllMultipleEdges` perhaps use this new function as its backend? Just an idea, haven't thought about it beyond writing this sentence.

So, this solves a different issue that the one solved by `DigraphRemoveAllMultipleEdges` it replaces multiple edges between the same nodes by single edges with a colour, and is used inside the canonical labelling stuff. I think it's also slower than `DigraphRemoveAllMultipleEdges`.

**Pull request merged**